{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import datetime\n",
    "plt.style.use('ggplot')\n",
    "sns.set_style(\"whitegrid\")\n",
    "%matplotlib inline\n",
    "%config InlineBackend.figure_format = 'svg'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_df(f, i):\n",
    "    items = f.split(\"/\")\n",
    "    binner, assembly, l = items[i-2], items[i-1], items[i]\n",
    "    _df = pd.read_csv(f, sep=\"\\t\", header=0, index_col=0)\n",
    "    data = {\"binner\": [binner]*_df.shape[0],\n",
    "            \"assembly\": [assembly]*_df.shape[0],\n",
    "            \"min_contig_length\": [l]*_df.shape[0]}\n",
    "    _df = pd.merge(pd.DataFrame(data, index=_df.index), _df, \n",
    "                   left_index=True, right_index=True)\n",
    "    return _df\n",
    "\n",
    "\n",
    "def read_summary_stats(f):\n",
    "    \"\"\"\n",
    "    Reads the basic summary statistics for bins\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(f, sep=\"\\t\", header=0)\n",
    "    value_vars = [\"Mbp\", \"GC\", \"contigs\", \"n50\"]\n",
    "    id_vars = [\"binner\", \"assembly\"]\n",
    "    # Use contig length for rows if doing multiple plotting\n",
    "    if len(df.min_contig_length.unique()) > 1:\n",
    "        id_vars.append(\"min_contig_length\")\n",
    "        row = \"min_contig_length\"\n",
    "    else:\n",
    "        row = None\n",
    "    # Melt the stats dataframe\n",
    "    dfm = pd.melt(df, id_vars=id_vars, value_vars=value_vars)\n",
    "    bin_counts = df.groupby([\"binner\",\"assembly\",\"min_contig_length\"]).count().loc[:,\"bp\"]\n",
    "    bin_counts = pd.DataFrame(bin_counts).rename(columns={'bp': 'bins'}).reset_index()\n",
    "    return dfm, bin_counts, row\n",
    "\n",
    "\n",
    "def read_checkm_stats(f):\n",
    "    \"\"\"\n",
    "    Reads extended statistics from checkm qa\n",
    "    \"\"\"\n",
    "    df = pd.read_csv(f, sep=\"\\t\", index_col=0, header=0)\n",
    "    df = df.assign(Purity = 100 - df.Contamination)\n",
    "    df = df.assign(Mbp = df[\"Genome size (bp)\"] / 1000000)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with PdfPages(snakemake.output[0]) as pdf:\n",
    "    #### Read the summary statistics ####\n",
    "    stats = [f for f in snakemake.input if os.path.basename(f) == \"binning_summary.tsv\"]\n",
    "    df, bin_counts, row = read_summary_stats(stats[0])\n",
    "    #### Plot the summary stats\n",
    "    plt.figure(figsize=(8, 8))\n",
    "    plt.title('Overall statistics')\n",
    "    sns.catplot(kind=\"strip\", col=\"variable\", hue=\"assembly\", y=\"value\", x=\"binner\", data=df, \n",
    "            hue_order=sorted(df.assembly.unique()), sharey=False, linewidth=.5, row=row)\n",
    "    pdf.savefig(bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    #### Plot number of bins\n",
    "    plt.rc('text', usetex=False)\n",
    "    plt.figure(figsize=(8, 8))\n",
    "    plt.title(\"Number of bins\")\n",
    "    sns.catplot(data=bin_counts, hue=\"assembly\", kind=\"bar\", x=\"binner\", y=\"bins\", \n",
    "            hue_order=sorted(bin_counts.assembly.unique()), col=\"min_contig_length\", sharey=True)\n",
    "    pdf.savefig(bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "    \n",
    "    #### Read checkm stats if available ####\n",
    "    checkm_stats = [f for f in snakemake.input if os.path.basename(f) == \"checkm.stats.tsv\"]\n",
    "    if len(checkm_stats) > 0:\n",
    "        checkm_df = read_checkm_stats(checkm_stats[0])\n",
    "        lengths = sorted(checkm_df.min_contig_length.unique())\n",
    "        plt.rc('text', usetex=False)\n",
    "        fig, a = plt.subplots(nrows=len(lengths), ncols=1, figsize=(8,6*len(lengths)))\n",
    "        if len(lengths) > 1:\n",
    "            axes = [axis for axis in a]\n",
    "        else:\n",
    "            axes = [a]\n",
    "        for i, l in enumerate(lengths):\n",
    "            ax = sns.scatterplot(data=checkm_df.loc[checkm_df.min_contig_length==l], y=\"Completeness\", x=\"Purity\", style=\"binner\", hue=\"assembly\", \n",
    "                           hue_order=sorted(checkm_df.assembly.unique()), size=\"Mbp\", linewidth=.5, ax=axes[i])\n",
    "            ax.set_title(\"min contig length={}\".format(l))\n",
    "            ax.legend(bbox_to_anchor=(1,1))\n",
    "            ax.set_xlabel(\"Purity (%)\")\n",
    "            ax.set_ylabel(\"Completeness (%)\")\n",
    "        pdf.savefig(fig, bbox_inches=\"tight\")\n",
    "        plt.close()\n",
    "    # Set the file's metadata via the PdfPages object:\n",
    "    d = pdf.infodict()\n",
    "    d['Title'] = 'Binning report for NBIS-meta workflow'\n",
    "    d['Author'] = 'NBIS'\n",
    "    d['CreationDate'] = datetime.datetime.today()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
